### ISSUE TITLE
ProCAT Protocol Stops Working

### ISSUE THREAD

**Issue by gdwaynewarner** [openstenoproject/plover#601]
Classification: Bug

Reproducibility: It has happened twice since installation, which was about two hours ago as I write this.
## Summary

With Plover set to the ProCAT protocol, I will be writing along, and suddenly Plover stops sending my text to whatever I'm writing in.
## Steps to Reproduce
1. I set Plover to use the new ProCAT protocol and write.
## Expected Results
1. This should happen: Plover should send my output to whatever I'm writing to for as long as I'm writing.
## Actual Results
1. Instead, this happen: Plover just stops sending what I'm writing to whatever I'm writing in at the time.  :(
   [plover_proCAT_error.txt](https://github.com/openstenoproject/plover/files/491974/plover_proCAT_error.txt)

**Comment by morinted** [comment]
I'm guessing this is happening because the `''` 'key' is getting sent and the keymap doesn't understand it.

Maybe instead of `steno_keys.append(STENO_KEY_CHART[i * 8 + j])` it would be best to do:

```
key = STENO_KEY_CHART[i * 8 + j]
if key:
steno_keys.append()
```

**Comment by gdwaynewarner** [comment]
You're welcome, Ted! Hope it helps to get it locked down!

**Comment by morinted** [comment]
Glen is still running into this issue. My suspicion is that we are running into a case where the machine is sending data right as our read timeout is ending, and we only pick up a partial piece of the packet. In this case, we should read the remainder of the packet to not become out of sync:

```python
            # Grab data from the serial port.
            raw = self.serial_port.read(BYTES_PER_STROKE)
            if not raw:
                continue
            if len(raw) < BYTES_PER_STROKE:  # If we only read a partial packet.
                raw += self.serial_port.read(BYTES_PER_STROKE - len(raw))
```

**Comment by benoit-pierre** [comment]
What's the status on this?

**Comment by morinted** [comment]
Still an issue, have not been able to get good debug data. I'm guessing
that the ignored bits contain some info indicating that the machine will
send info, maybe a timestamp?

**Comment by benoit-pierre** [comment]
Should be fixed by #866.
